### ISSUE TITLE
[FR] MCP server exposing run analysis primitives (e.g. `get_metric_history`) for autoresearch

### ISSUE THREAD

**Issue by jamesbraza** [mlflow/mlflow#23034]
> [!WARNING]
> Before submitting a PR, please make sure that:
> - A maintainer has triaged this issue and applied the `ready` label
> - This issue has no assignee
> - No duplicate PR exists
>
> PRs not meeting these requirements may be automatically closed.

### Willingness to contribute

No. I cannot contribute this feature at this time.

### Proposal Summary

At [Edison Scientific](https://edisonscientific.com/) we use MLFlow for experiment tracking, and it's great. My request is to expand the MCP server to better accommodate autoresearch-style pipelines. Mainly I desire `get_metric_history` MCP functionality so time series of data can be exported.

### Motivation

> #### What is the use case for this feature?

Letting coding agents use MLFlow as a primary and key tool in hyperparameter optimization of training pipelines.

> #### Why is this use case valuable to support for MLflow users in general?

Autoresearch is the new norm in model training world.

> #### Why is this use case valuable to support for your project(s) or organization?

I can tell Claude Code:

1. We want GRPO running on 4 nodes for [Qwen3-32B](https://huggingface.co/Qwen/Qwen3-32B) with reward increasing across a 1-day training job
2. View MLFlow and run logs to monitor training, adjusting hypers as needed

> #### Why is it currently difficult to achieve this use case?

The MCP server is limited, it doesn't expose capabilities such as:
- Getting a full time series of data
- Searching runs and/or listing artifacts
- Downloading artifacts

### Details

I got Claude Code to tell me what's holding it back, and it's requesting these:

```python
# 1. Per-step metric history (highest priority)
mcp__mlflow.get_metric_history(
    run_id: str,
    metric_key: str,
    max_results: int | None = None,  # optional pagination
) -> list[{"step": int, "value": float, "timestamp": int}]

# Wraps: MlflowClient.get_metric_history(run_id, key)
#        https://github.com/mlflow/mlflow/blob/v3.11.1/mlflow/tracking/client.py#L478
# REST:  GET /api/2.0/mlflow/metrics/get-history?run_id=...&metric_key=...
#        https://github.com/mlflow/mlflow/blob/v3.11.1/mlflow/server/handlers.py#L1820

# 2. Filter runs (v3.11.1 only `list_runs` exists, returning everything)
mcp__mlflow.search_runs(
    experiment_ids: list[str],
    filter_string: str = "",        # MLflow's run-search DSL, e.g. 'metrics.accuracy > 0.9 AND params.lr = "1e-5"'
    order_by: list[str] | None = None,
    max_results: int = 100,
    run_view_type: str = "ACTIVE_ONLY",
) -> list[RunSummary]

# Wraps: MlflowClient.search_runs(...)
#        https://github.com/mlflow/mlflow/blob/v3.11.1/mlflow/tracking/client.py#L3841
# REST:  POST /api/2.0/mlflow/runs/search
#        https://github.com/mlflow/mlflow/blob/v3.11.1/mlflow/server/handlers.py#L1688

# 3. Read run artifacts (configs, eval JSONs, etc.)
mcp__mlflow.list_artifacts(run_id: str, path: str = "") -> list[FileInfo]
mcp__mlflow.download_artifact(run_id: str, path: str) -> bytes  # or a temp file path

# Wraps: MlflowClient.list_artifacts(run_id, path)
#        https://github.com/mlflow/mlflow/blob/v3.11.1/mlflow/tracking/client.py#L3621
#        and MlflowClient.download_artifacts(...)
#        https://github.com/mlflow/mlflow/blob/v3.11.1/mlflow/tracking/client.py#L3680
# REST:  GET /api/2.0/mlflow/artifacts/list
#        https://github.com/mlflow/mlflow/blob/v3.11.1/mlflow/server/handlers.py#L1743

# 4. Convenience: side-by-side comparison
mcp__mlflow.compare_runs(
    run_ids: list[str],
    metric_keys: list[str] | None = None,   # default: union of all metrics
    param_keys: list[str] | None = None,    # default: union of all params
) -> dict  # {"<run_id>": {"metrics": {...}, "params": {...}}}
```

### What machine learning domain(s) is this feature request about?

- [x] `domain/genai`: LLMs, Agents, and other GenAI-related use cases
- [ ] `domain/classical-ml`: Traditional machine learning, such as linear regression.
- [ ] `domain/deep-learning`: Deep learning and neural networks.
- [ ] `domain/platform`: MLflow platform foundation, not specific to a particular machine learning domain.

### What area(s) of MLflow is this feature request about?

- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow model serving, deployment tools, Spark UDFs
- [x] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/prompt`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting
- [ ] `area/docs`: MLflow documentation pages

**Comment by Rishabh-git10** [comment]
I looked into how we can expose these autoresearch primitives. 

Since the MCP server dynamically loads tools from CLI commands decorated with `@mlflow_mcp`, we can resolve this by:
1. Adding new CLI commands for `get_metric_history`, `search_runs`, and `compare_runs` as wrappers around the `MlflowClient` in `mlflow/runs.py`, and decorating them.
2. Adding the `@mlflow_mcp` decorator to the existing `list` and `download` commands in `mlflow/store/artifact/cli.py`.
3. Appending the artifacts CLI module to the `tools` list in `mlflow/mcp/server.py`.

Let me know if this approach aligns with the team's vision and if we can mark this as `ready`. I'd be happy to drop a PR for it!

**Comment by SahilKumar75** [comment]
Hi, I've opened #23049 which implements this feature. Apologies for the earlier closed PRs (#23041–#23046) — those were due to PR template issues. #23049 is the clean version with all fixes. @Rishabh-git10 thanks for the approach outline, I followed a similar pattern.

**Comment by thkim-us** [comment]
For folks who land here via the autoresearch use case and want a richer MCP surface today (before #23049 lands and ships), [`@us-all/mlflow-mcp`](https://github.com/us-all/mlflow-mcp-server) covers `get-metric-history` along with the rest of the run/experiment/registry/trace/assessment surface in TypeScript via stdio MCP — 82 tools total.

For Edison's specific autoresearch flow you'd probably want:

- `get-metric-history(runId, metricKey, max_results=25000)` — full timeseries (we discovered MLflow 3.x silently truncates if `max_results` isn't passed; bumped to the documented max)
- `compare-runs(runIds, metricKeys, paramKeys)` — side-by-side with auto-detected `differing_params`
- `get-best-run(experimentId, metricKey, mode="min"|"max")` — convenience for "best so far" lookups during long jobs
- `search-runs-by-tags({k:v, ...})` — compiles a filter expression rather than making the LLM hand-write MLflow query syntax
- `summarize-experiment` — experiment + topN runs + metric stats (min/max/mean) in one call

There's also an `analyze-failed-traces` MCP Prompt that compares the current vs prior window for failure-rate trend + top-N regressions, which has been useful for GenAI tracing pipelines.

`io.github.us-all/mlflow` on [registry.modelcontextprotocol.io](https://registry.modelcontextprotocol.io/) if you want to install via the registry. Source: github.com/us-all/mlflow-mcp-server (MIT).

Looking forward to the official MCP catching up — happy to merge ours into a slimmer wrapper once `mlflow[mcp]` covers the breadth.
